chain/epoch-manager/src/reward_calculator.rs

```diff
@@ -88,23 +88,88 @@ impl RewardCalculator {
+                    ),
+                    // Validator produced blocks and chunks, but not endorsements
+                    (expected_blocks_produced, expected_chunks_produced, 0) => {
+                        let numer = U256::from(
+                            stats.block_stats.produced * expected_chunks_produced
+                                + stats.chunk_stats.produced() * expected_blocks_produced,
+                        );
+                        let denom =
+                            U256::from(2 * expected_chunks_produced * expected_blocks_produced);
+                        (numer, denom)
+                    }
+                    // Validator produced chunks and endorsements, but not blocks
+                    (0, expected_chunks_produced, expected_endorsements) => {
+                        let endorsement_stats = stats.chunk_stats.endorsement_stats();
+                        let numer = U256::from(
+                            endorsement_stats.produced * expected_chunks_produced
+                                + stats.chunk_stats.produced() * expected_endorsements,
+                        );
+                        let denom =
+                            U256::from(2 * expected_chunks_produced * expected_endorsements);
+                        (numer, denom)
+                    }
+                    // Validator produced blocks and endorsements, but not chunks
+                    (expected_blocks_produced, 0, expected_endorsements) => {
+                        let endorsement_stats = stats.chunk_stats.endorsement_stats();
+                        let numer = U256::from(
+                            endorsement_stats.produced * expected_blocks_produced
+                                + stats.block_stats.produced * expected_endorsements,
+                        );
+                        let denom =
+                            U256::from(2 * expected_blocks_produced * expected_endorsements);
+                        (numer, denom)
+                    }
+                    // Validator did all the things
+                    (expected_blocks_produced, expected_chunks_produced, expected_endorsements) => {
+                        let produced_blocks = stats.block_stats.produced;
+                        let produced_chunks = stats.chunk_stats.produced();
+                        let produced_endorsements = stats.chunk_stats.endorsement_stats().produced;
+
+                        let numer = U256::from(
+                            produced_blocks * expected_chunks_produced * expected_endorsements
+                                + produced_chunks
+                                    * expected_blocks_produced
+                                    * expected_endorsements
+                                + produced_endorsements
+                                    * expected_blocks_produced
+                                    * expected_chunks_produced,
+                        );
```

Comment:
As we are increasing number of factors, can we add some comments why overflows don't happen at any point? We recently hit a crash because of that in statelessnet so we need to be careful about that.

Reply:
There is `test_reward_no_overflow` which checks the calculation still works with the anticipated maximum values.